source/core_json.c

```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
+
+    skipSpace( buf, &i, max );
+
+    if( ( i < max ) && ( buf[ i ] == '[' ) )
+    {
+        i++;
+        skipSpace( buf, &i, max );
+
+        while( i < max )
         {
-            value++;
-            valueLength -= 2U;
+            if( nextValue( buf, &i, max, &value, &valueLength ) != true )
+            {
+                break;
+            }
+
+            if( j == queryIndex )
+            {
+                ret = true;
+                break;
+            }
+
+            if( skipSpaceAndComma( buf, &i, max ) != true )
+            {
+                break;
+            }
+
+            j++;
         }
+    }
 
+    if( ret == true )
+    {
         *outValue = &buf[ value ];
         *outValueLength = valueLength;
     }
-    else if( ret == JSONPartial )
+
+    return ret;
+}
+
+/**
+ * @brief Advance buffer index beyond a query key subpart.
+ *
+ * @param[in] buf  The buffer to parse.
+ * @param[in,out] start  The index at which to begin.
+ * @param[in] max  The size of the buffer.
+ *
+ * @return true if a valid string was present;
+ * false otherwise.
+ */
+#define JSON_QUERY_KEY_SEPARATOR    '.'
+#define isSeparator_( x )    ( ( x ) == JSON_QUERY_KEY_SEPARATOR )
+static bool_ skipQueryKey( const char * buf,
+                           size_t * start,
+                           size_t max,
+                           size_t * outLength )
+{
+    bool_ ret = false;
+    size_t i;
+
+    assert( ( buf != NULL ) && ( start != NULL ) && ( outLength != NULL ) );
+    assert( max > 0U );
+
+    i = *start;
+
+    while( ( i < max ) &&
+           !isSeparator_( buf[ i ] ) &&
+           !isSquareOpen_( buf[ i ] ) )
+    {
+        i++;
+    }
+
+    if( i > *start )
+    {
+        ret = true;
+        *outLength = i - *start;
+        *start = i;
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Handle a nested search by iterating over the parts of the queryKey.
+ *
+ * @param[in] buf  The buffer to search.
+ * @param[in] max  size of the buffer.
+ * @param[in] queryKey  The key to search for.
+ * @param[in] queryKeyLength  Length of the key.
+ * @param[out] outValue  A pointer to receive the address of the value found.
+ * @param[out] outValueLength  A pointer to receive the length of the value found.
+ *
+ * @return #JSONSuccess if the queryKey is found and the value output;
+ * #JSONBadParameter if the queryKey is empty, or any subpart is empty,
+ * or an index is too large to convert;
+ * #JSONNotFound if the queryKey is NOT found.
+ *
+ * @note Parsing stops upon finding a match.
+ */
+static JSONStatus_t multiSearch( char * buf,
+                                 size_t max,
+                                 const char * queryKey,
+                                 size_t queryKeyLength,
+                                 char ** outValue,
+                                 size_t * outValueLength )
+{
+    JSONStatus_t ret = JSONSuccess;
+    size_t i = 0, start = 0, keyLength = 0;
+    char * p = buf;
```

Comment:
We probably should use more expressive variable names that convey the meaning of the variable. Referring to `i` and `p` variable names

Reply:
These are idiomatic usages, and passed muster with previous reviewers.  `i` is used in the same fashion in almost every function.  We can discuss whether idioms are a good or bad thing, but I'd rather not do so in the context of this PR, given our time considerations.